src/states/common/base.rs

```diff
@@ -85,4 +87,43 @@ pub trait Base: Debug {
             // return self.handle_lost_peer(*pub_info).map(|_| Err(err.into()));
         }
     }
+
+    fn sign_direct_message(
+        &self,
+        direct_message: &DirectMessage,
+    ) -> Result<sign::Signature, RoutingError> {
+        let serialised_content = match serialisation::serialise(direct_message) {
```

Comment:
can we make this serialisation a function to reduce code duplication?

Reply:
That seems a bit like overkill to me since the two places it'd be used need to return different things.  We actually end up with more code:
```rust
    fn sign_direct_message(
        &self,
        direct_message: &DirectMessage,
    ) -> Result<sign::Signature, RoutingError> {
        let serialised_content = self.serialise_direct_message(direct_message)?;
        Ok(sign::sign_detached(
            &serialised_content,
            self.full_info().secret_sign_key(),
        ))
    }

    fn verify_direct_message(
        &self,
        direct_message: &DirectMessage,
        signature: &sign::Signature,
        pub_info: &PublicInfo,
    ) -> bool {
        let serialised_content = match self.serialise_direct_message(direct_message) {
            Ok(result) => result,
            Err(_) => {
                return false;
            }
        };

        if !sign::verify_detached(signature, &serialised_content, pub_info.sign_key()) {
            info!("Direct message has an invalid signature");
            return false;
        }

        true
    }

    fn serialise_direct_message(
        &self,
        direct_message: &DirectMessage,
    ) -> Result<Vec<u8>, RoutingError> {
        serialisation::serialise(direct_message).map_err(|error| {
            error!("Failed to serialise direct message: {:?}", error);
            error.into()
        })
    }
```

Reply:
3 more lines VS. de-duplication, hard to decide   :)

btw. we do have some places in code that carrying out similar serialisation result check then log, such as:
```rust
            let mut to_sign = match serialisation::serialise(&old_and_new_pub_infos) {
                Ok(result) => result,
                Err(error) => {
                    error!("Failed to serialise public IDs: {:?}", error);
                    return;
                }
            };
```

just wondering, will it be worth to make it a helper macro.
it doesn't need to be in this PR, can be done in a separate one.

Reply:
Hmm, I try and avoid macros where possible.  We can maybe try it later to see how it looks?